integration-libs/opps/coupon-codes/http-interceptors/occ-opps-coupon.interceptor.ts

```diff
@@ -0,0 +1,91 @@
+/*
+ * SPDX-FileCopyrightText: 2024 SAP Spartacus team <spartacus-team@sap.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {
+  HttpEvent,
+  HttpHandler,
+  HttpInterceptor,
+  HttpRequest,
+} from '@angular/common/http';
+import { Injectable, inject, isDevMode } from '@angular/core';
+import { Observable } from 'rxjs';
+import { LoggerService, OccEndpointsService, WindowRef } from '@spartacus/core';
+import { OppsConfig } from '../config/opps-config';
+import {
+  OPPS_COUPON_LOCAL_STORAGE_KEY,
+  OPPS_COUPON_URL_QUERY_PARAM,
+} from '../config';
+
+@Injectable({ providedIn: 'root' })
+export class OccOppsCouponInterceptor implements HttpInterceptor {
+  private oppsCoupon?: string | null;
+  private requestHeader?: string;
+
+  protected logger = inject(LoggerService);
+
+  protected config = inject(OppsConfig);
+  protected occEndpoints = inject(OccEndpointsService);
+  protected winRef = inject(WindowRef);
+
+  constructor() {
+    this.initialize();
+  }
+
+  /**
+   * Fetched the OPPS coupon codes from URL query parameter and saves it into
+   * browser local storage
+   */
+  protected initialize() {
+    const url = this.winRef.location.href ?? '';
+    const queryParams = new URLSearchParams(url.substring(url.indexOf('?')));
+    this.oppsCoupon = queryParams.get(OPPS_COUPON_URL_QUERY_PARAM);
+    if (this.oppsCoupon) {
+      this.winRef.localStorage?.setItem(
+        OPPS_COUPON_LOCAL_STORAGE_KEY,
+        this.oppsCoupon
+      );
+    } else {
+      this.oppsCoupon = this.winRef.localStorage?.getItem(
+        OPPS_COUPON_LOCAL_STORAGE_KEY
+      );
+    }
+    if (this.winRef.isBrowser()) {
+      if (!this.config.opps?.couponcodes?.httpHeaderName && isDevMode()) {
+        this.logger.warn(
+          `There is no httpHeaderName configured for OPPS Coupon`
+        );
+      }
+      this.requestHeader =
+        this.config.opps?.couponcodes?.httpHeaderName?.toLowerCase?.();
+    }
+  }
```

Comment:
Perhaps you might want to also wrap the logic responsible for saving/reading from local storage to some simple service, e.g. `OppsCouponCodesService` with 2 methods `set(value: string)` and `get(): string` - which under the hood communicate with localStorage

Reply:
I guess we should implement some logic for clearing the saved value, e.g. on user LOGOUT event. Or at least we should consult it with the product owner of the OPPS feature.

Otherwise if an user logs out, then another user of this same browser will implicitly "inherit" the coupon codes - which might be potentially against the business rules.

Reply:
Hi Kris, If we use LOCATION_INITIALIZED_MULTI to capture coupon codes, factory method will be called only once and not the second time, when user launches a another url with a different coupon code -> am i missing something where?